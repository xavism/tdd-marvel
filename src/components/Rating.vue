<template>
  <div class="star-rating">
    <fieldset>
      <input type="radio" id="star5" name="rating" value="5" v-model="rating">
      <label for="star5" title="Outstanding">5 stars</label>
      <input type="radio" id="star4" name="rating" value="4" v-model="rating">
      <label for="star4" title="Very Good">4 stars</label>
      <input type="radio" id="star3" name="rating" value="3" v-model="rating">
      <label for="star3" title="Good">3 stars</label>
      <input type="radio" id="star2" name="rating" value="2" v-model="rating">
      <label for="star2" title="Poor">2 stars</label>
      <input type="radio" id="star1" name="rating" value="1" v-model="rating">
      <label for="star1" title="Very Poor">1 star</label>
    </fieldset>
  </div>
</template>

<script>
export default {
  name: 'Rating',
  props: {
    rating: {
      type: Number,
      required: true,
    },
  },
  watch: {
    rating: function() {
      console.log(`EMITING ${this.rating}`);
      this.$emit('starsChanges', this.rating);
    }
  },
};
</script>

<style scoped>
.star-rating {
  font-family: "FontAwesome";
  margin: 50px auto;
}
.star-rating > fieldset {
  border: none;
  display: inline-block;
}
.star-rating > fieldset:not(:checked) > input {
  position: absolute;
  top: -9999px;
  clip: rect(0, 0, 0, 0);
}
.star-rating > fieldset:not(:checked) > label {
  float: right;
  width: 1em;
  padding: 0 0.05em;
  overflow: hidden;
  white-space: nowrap;
  cursor: pointer;
  font-size: 200%;
  color: #16a085;
}
.star-rating > fieldset:not(:checked) > label:before {
  content: "\f006  ";
}
.star-rating > fieldset:not(:checked) > label:hover,
.star-rating > fieldset:not(:checked) > label:hover ~ label {
  color: #1abc9c;
  text-shadow: 0 0 3px #1abc9c;
}
.star-rating > fieldset:not(:checked) > label:hover:before,
.star-rating > fieldset:not(:checked) > label:hover ~ label:before {
  content: "\f005  ";
}
.star-rating > fieldset > input:checked ~ label:before {
  content: "\f005  ";
}
.star-rating > fieldset > label:active {
  position: relative;
  top: 2px;
}
body {
  background: #262829;
  color: #95a5a6;
  font-family: "Raleway";
  text-align: center;
}
body p {
  font-size: 1.6em;
  margin: auto;
  width: 80%;
}
body a {
  color: #16a085;
}
</style>
