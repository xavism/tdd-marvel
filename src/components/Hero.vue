<template>
    <div>
        <img :src="imagePath" alt="hero Image" class="image-avatar">
        <h1 class="hero-name">{{hero.name}}</h1>
        <Rating :rating="hero.rating"/>
    </div>
</template>

<script>
import Rating from '@/components/Rating';

export default {
  name: 'Hero',
  components: {
    Rating,
  },
  props: {
    hero: {
      type: Object,
      required: true,
    },
  },
  computed: {
    imagePath() {
      return `${this.hero.thumbnail.path}.${this.hero.thumbnail.extension}`;
    },
  },
};
</script>
