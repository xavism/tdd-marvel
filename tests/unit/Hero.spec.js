import { shallowMount } from '@vue/test-utils';
import Hero from '@/components/Hero';
import Rating from '@/components/Rating';
import heroFixture from './fixtures/hero';

describe('Hero', () => {
  let props;

  const build = () => {
    const options = { propsData: props };
    const wrapper = shallowMount(Hero, options);
    return {
      wrapper,
      avatar: () => wrapper.find('.image-avatar'),
      name: () => wrapper.find('.hero-name'),
      rating: () => wrapper.find(Rating),
    };
  };

  beforeEach(() => {
    props = {
      hero: heroFixture,
    };
  });

  it('renders the component', () => {
    // arrange
    const { wrapper } = build();
    // assert
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders main components correctly', () => {
    // arrange
    const { avatar, name, rating } = build();
    // assert
    expect(avatar().exists()).toBe(true);
    const thumbnail = `${props.hero.thumbnail.path}.${props.hero.thumbnail.extension}`;
    expect(avatar().attributes().src).toBe(thumbnail);
    expect(rating().exists()).toBe(true);
    expect(name().exists()).toBe(true);
    expect(name().text()).toBe(props.hero.name);
  });
});
