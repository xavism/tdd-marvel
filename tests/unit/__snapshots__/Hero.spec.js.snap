// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero renders the component 1`] = `
<div><img src="http://i.annihil.us/u/prod/marvel/i/mg/d/d0/5269657a74350.jpg" alt="hero Image" class="image-avatar">
  <h1 class="hero-name">Thor</h1>
  <rating-stub rating="2"></rating-stub>
</div>
`;
